src/main/java/picard/illumina/ExtractBarcodesProgram.java

```diff
@@ -0,0 +1,312 @@
+        }
+        metrics.addMetric(noMatchMetric);
+        metrics.write(METRICS_FILE);
+    }
+
+    public static void finalizeMetrics(final Map<String, BarcodeMetric> barcodeToMetrics,
+                                       final BarcodeMetric noMatchMetric) {
+        // Finish metrics tallying.
+        long totalReads = noMatchMetric.READS;
+        long totalPfReads = noMatchMetric.PF_READS;
+        long totalPfReadsAssigned = 0;
+        for (final BarcodeMetric barcodeMetric : barcodeToMetrics.values()) {
+            totalReads += barcodeMetric.READS;
+            totalPfReads += barcodeMetric.PF_READS;
+            totalPfReadsAssigned += barcodeMetric.PF_READS;
+        }
+
+        if (totalReads > 0) {
+            noMatchMetric.PCT_MATCHES = noMatchMetric.READS / (double) totalReads;
+            double bestPctOfAllBarcodeMatches = 0;
+            for (final BarcodeMetric barcodeMetric : barcodeToMetrics.values()) {
+                barcodeMetric.PCT_MATCHES = barcodeMetric.READS / (double) totalReads;
+                if (barcodeMetric.PCT_MATCHES > bestPctOfAllBarcodeMatches) {
+                    bestPctOfAllBarcodeMatches = barcodeMetric.PCT_MATCHES;
+                }
+            }
+            if (bestPctOfAllBarcodeMatches > 0) {
+                noMatchMetric.RATIO_THIS_BARCODE_TO_BEST_BARCODE_PCT =
+                        noMatchMetric.PCT_MATCHES / bestPctOfAllBarcodeMatches;
+                for (final BarcodeMetric barcodeMetric : barcodeToMetrics.values()) {
+                    barcodeMetric.RATIO_THIS_BARCODE_TO_BEST_BARCODE_PCT =
+                            barcodeMetric.PCT_MATCHES / bestPctOfAllBarcodeMatches;
+                }
+            }
+        }
+
+        if (totalPfReads > 0) {
+            noMatchMetric.PF_PCT_MATCHES = noMatchMetric.PF_READS / (double) totalPfReads;
+            double bestPctOfAllBarcodeMatches = 0;
+            for (final BarcodeMetric barcodeMetric : barcodeToMetrics.values()) {
+                barcodeMetric.PF_PCT_MATCHES = barcodeMetric.PF_READS / (double) totalPfReads;
+                if (barcodeMetric.PF_PCT_MATCHES > bestPctOfAllBarcodeMatches) {
+                    bestPctOfAllBarcodeMatches = barcodeMetric.PF_PCT_MATCHES;
+                }
+            }
+            if (bestPctOfAllBarcodeMatches > 0) {
+                noMatchMetric.PF_RATIO_THIS_BARCODE_TO_BEST_BARCODE_PCT =
+                        noMatchMetric.PF_PCT_MATCHES / bestPctOfAllBarcodeMatches;
+                for (final BarcodeMetric barcodeMetric : barcodeToMetrics.values()) {
+                    barcodeMetric.PF_RATIO_THIS_BARCODE_TO_BEST_BARCODE_PCT =
+                            barcodeMetric.PF_PCT_MATCHES / bestPctOfAllBarcodeMatches;
+                }
+            }
+        }
+
+        // Calculate the normalized matches
+        if (totalPfReadsAssigned > 0) {
+            final double mean = (double) totalPfReadsAssigned / (double) barcodeToMetrics.values().size();
```

Comment:
Can barcodeToMetrics ever be empty?

Reply:
If it was empty than it wouldn't pass command line validation

```
            if (barcodeToMetrics.keySet().isEmpty()) {
                messages.add("No barcodes have been specified.");
            }
```